controller/src/results.rs

```diff
@@ -247,6 +248,62 @@ impl<P> ModuleResult<P> {
     }
 }
 
+impl<P: Clone> ModuleResult<P> {
+    /// Merge the successful and error data from one result into `self`.
+    ///
+    /// This will take all the successful module IDs and data, and merge them
+    /// into `self`'s success. Any failures in `other` will be merged into
+    /// `self.failures`.
+    ///
+    /// This returns `None` if any of the following occur:
+    ///
+    /// - `other.modules` and `self.modules` overlap. In this case, it is not
+    /// possible to build a consistent result, since we have two values for the
+    /// successful data.
+    /// - `other.failures` and `self.failures` overlap, for the same reason.
+    ///
+    /// Note that if a module exists in `other.failures.modules` and
+    /// `self.modules`, then it is _removed_ from `self.modules` and added to
+    /// the failures. The converse is _not_ true: if a module exists in
+    /// `self.failures.modules` and `other.modules` it is _not_ removed. That
+    /// is, failures are "sticky" in `self`.
+    pub fn merge(&self, other: &Self) -> Option<Self> {
+        // The intersection of the successful module IDs should be empty.
+        if (self.modules.0 & other.modules.0) != 0 {
```

Comment:
Nit: this suggests that we should implement `BitAnd` and `is_empty` for `ModuleId`, so this could be written
```rust
if !(self.modules & other.modules).is_empty() {
```

Aaron added a bunch of `impl Bit*` in Hubris [here](https://github.com/oxidecomputer/hubris/blob/10fb31352fd25a6b6dba01fe8102cf080482708e/drv/sidecar-front-io/src/transceivers.rs#L197-L223), which helped make the code more obvious.

Reply:
Yeah, I've put that off for a while, relying on explicit functions. We've not really needed them to date, but I can add them now.

Reply:
Done in `3ddd1fd`